### PR TITLE
Feature/use sf authentication session

### PR DESCRIPTION
**Pull request by TroubleMakerBen**

### CODE REVIEW

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 1]
```diff
@@ -446,6 +454,35 @@ - (nullable NSURL *)buildTokenRequestURLWithAppSecret:(NSString *)appSecret rele
...
+  if (@available(iOS 11.0, *)) {
+    SFAuthenticationSession *session = [[clazz alloc] initWithURL:url callbackURLScheme:callbackUrlScheme completionHandler:^(NSURL *callbackUrl, NSError *error) {
+      
+      self.authenticationSession = nil;
+      MSLogDebug([MSDistribute logTag], @"Called %@ with errror: %@", callbackUrl, error.localizedDescription);
+      
+      if(error.code == SFAuthenticationErrorCanceledLogin) {
+        MSLogError([MSDistribute logTag], @"Authentication session was cancelled by user or failed.");
+      }
+
```
Comment (MatkovIvan) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: > No newlines within methods except in front of a comment

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 2]
```diff
@@ -220,17 +220,25 @@ - (void)requestInstallInformationWith:(NSString *)releaseHash {
...
-       * TODO Checking operating system version is a workaround for iOS 11 where SFSafariViewController can't read Safari's cookies.
-       * Revert this change when SFAuthenticationSession will be ready.
-       */
       if (clazz && ![NSProcessInfo.processInfo isOperatingSystemAtLeastVersion:(NSOperatingSystemVersion){11, 0, 0}]) {
 
-        // Manipulate App UI on the main queue.
-        dispatch_async(dispatch_get_main_queue(), ^{
-          [self openURLInEmbeddedSafari:url fromClass:clazz];
-        });
+        // iOS 11
```
Comment (jaelim-ms) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]:
```
// Manipulate App UI on the main queue.
Class authClazz = [SFAuthenticationSession class];
if(authClazz) {
           
  // iOS 11
  dispatch_async(dispatch_get_main_queue(), ^{
    [self openURLInAuthenticationSessionWith:url fromClass:authClazz];
  });
} else {

  // iOS 9 and 10
  dispatch_async(dispatch_get_main_queue(), ^{
    [self openURLInEmbeddedSafari:url fromClass:clazz];
  });
} 
```
You can relocate the comment in this change like above.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 3]
```diff
@@ -446,6 +454,35 @@ - (nullable NSURL *)buildTokenRequestURLWithAppSecret:(NSString *)appSecret rele
...
+- (void)openURLInAuthenticationSessionWith:(NSURL *)url fromClass:(Class)clazz {
+  MSLogDebug([MSDistribute logTag], @"Using SFAuthenticationSession to open URL: %@", url);
+  
+  NSString *callbackUrlScheme = [NSString stringWithFormat:kMSDefaultCustomSchemeFormat, self.appSecret];
+  
+  if (@available(iOS 11.0, *)) {
+    SFAuthenticationSession *session = [[clazz alloc] initWithURL:url callbackURLScheme:callbackUrlScheme completionHandler:^(NSURL *callbackUrl, NSError *error) {
+      
+      self.authenticationSession = nil;
+      MSLogDebug([MSDistribute logTag], @"Called %@ with errror: %@", callbackUrl, error.localizedDescription);
```
Comment (jaelim-ms) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Can you check and print the log only if `error != nil`? If you want callbackUrl to be in the device log, please add one more log with callback Url.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 4]
```diff
@@ -446,6 +454,35 @@ - (nullable NSURL *)buildTokenRequestURLWithAppSecret:(NSString *)appSecret rele
...
+    self.authenticationSession = session;
+
+    BOOL success = [session start];
+    if(success) {
+      MSLogDebug([MSDistribute logTag], @"Authentication Session Started, showing confirmation dialog");
+    }
+  }
+}
+
+// TODO: rename method
```
Comment (jaelim-ms) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Let's rename now! What's your suggestion?

Comment (jaelim-ms) [reply MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: `openURLInSafariViewController`??

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 5]
```diff
@@ -556,13 +593,13 @@ - (BOOL)handleUpdate:(MSReleaseDetails *)details {
 }
 
 - (BOOL)checkForUpdatesAllowed {
+
```
Comment (jaelim-ms) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Unnecessary changes in this method. Please revert.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistributePrivate.h [hunk 6]
```diff
@@ -101,6 +107,13 @@ static NSString *const kMSDistributionGroupIdKey = @"MSDistributionGroupId";
  */
 @property(nonatomic) id<MSAppDelegate> appDelegate;
 
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= 110000
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wpartial-availability"
+@property(nullable, nonatomic) SFAuthenticationSession *authenticationSession;
+#pragma clang diagnostic pop
+#endif // __IPHONE_OS_VERSION_MAX_ALLOWED >= 110000
```
Comment (jaelim-ms) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistributePrivate.h]: This `#if` - `#endif` is not quite hard to read. Let's remove the comment at the end of the line.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m (hunk 4 above)
Comment (TroubleMakerBen) [reply MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: I've renamed it to`openURLInSafariViewControllerWithUrl:fromClass:` to make it work with the conventions...I ommited passing in the instance of SFSafariViewController though as we don't need that and it's an internal API.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 7]
```diff
@@ -214,23 +214,32 @@ - (void)requestInstallInformationWith:(NSString *)releaseHash {
...
+        if (authClazz) {
+
+          // Manipulate App UI on the main queue.
+          dispatch_async(dispatch_get_main_queue(), ^{
+            [self openURLInAuthenticationSessionWith:url fromClass:authClazz];
+          });
+        } else {
+          // iOS 9 and 10
+
+          // Manipulate App UI on the main queue.
```
Comment (jaelim-ms) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Can you just combine the two comments like `Manipulate App UI on the main queue on iOS 11`, `Manipulate App UI on the main queue on iOS 9 and 10`?
Or, you can just remove `// Manipulate App UI on the main queue.` since it can be easily realized that it is on main queue.
`// iOS 11` should be in if statement, `// iOS9 and 10` should have a line break above.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 8]
```diff
@@ -214,23 +214,32 @@ - (void)requestInstallInformationWith:(NSString *)releaseHash {
...
  */
 #pragma clang diagnostic push
 #pragma clang diagnostic ignored "-Wpartial-availability"
       Class clazz = [SFSafariViewController class];
-#pragma clang diagnostic pop
-      /*
-       * TODO Checking operating system version is a workaround for iOS 11 where SFSafariViewController can't read Safari's cookies.
-       * Revert this change when SFAuthenticationSession will be ready.
-       */
       if (clazz && ![NSProcessInfo.processInfo isOperatingSystemAtLeastVersion:(NSOperatingSystemVersion){11, 0, 0}]) {
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: I think there is two possibility to get rid of the heavy version check logic introduced in the workaround.
Either use feature availability. Something like
```
if SFAuthenticationSession class exists
  ...
else if SFSafariViewController class exists
  ...
else
  ...
```
Or the new `@available`.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 9]
```diff
@@ -446,7 +455,38 @@ - (nullable NSURL *)buildTokenRequestURLWithAppSecret:(NSString *)appSecret rele
   return components.URL;
 }
 
-- (void)openURLInEmbeddedSafari:(NSURL *)url fromClass:(Class)clazz {
+- (void)openURLInAuthenticationSessionWith:(NSURL *)url fromClass:(Class)clazz {
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Should add comments after newline or delete the newline ;)

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 10]
```diff
@@ -446,7 +455,38 @@ - (nullable NSURL *)buildTokenRequestURLWithAppSecret:(NSString *)appSecret rele
...
+  MSLogDebug([MSDistribute logTag], @"Using SFAuthenticationSession to open URL: %@", url);
+
+  NSString *callbackUrlScheme = [NSString stringWithFormat:kMSDefaultCustomSchemeFormat, self.appSecret];
+
+  if (@available(iOS 11.0, *)) {
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Neat :)

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 11]
```diff
@@ -446,7 +450,38 @@ - (nullable NSURL *)buildTokenRequestURLWithAppSecret:(NSString *)appSecret rele
   return components.URL;
 }
 
-- (void)openURLInEmbeddedSafari:(NSURL *)url fromClass:(Class)clazz {
+- (void)openURLInAuthenticationSessionWith:(NSURL *)url fromClass:(Class)clazz {
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Not on you but I realized seeing `openURLInAuthenticationSessionWith:fromClass:` that we should remove the last `Url` from `openURLInSafariViewControllerWithUrl:fromClass` :)

Comment (TroubleMakerBen) [reply MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: I'll do that!

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m (hunk 10 above)
Comment (TroubleMakerBen) [reply MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Problem is that this requires Xcode 9 and Xcode 9 breaks our tests for MSDistribute (not related to this PR)

File: MobileCenter/MobileCenter.xcodeproj/project.pbxproj [hunk 12]
```diff
@@ -2058,13 +2058,15 @@
 					"DEBUG=1",
 					"$(inherited)",
 				);
+				IPHONEOS_DEPLOYMENT_TARGET = 8.0;
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter.xcodeproj/project.pbxproj]: Why do you need this in project.pbxproj? It is already in Global.xcconfig.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 13]
```diff
@@ -214,28 +213,32 @@ - (void)requestInstallInformationWith:(NSString *)releaseHash {
...
+        } else {
+
+          // iOS 9 and 10
+          Class clazz = [SFSafariViewController class];
+          dispatch_async(dispatch_get_main_queue(), ^{
+            [self openURLInSafariViewControllerWithUrl:url fromClass:clazz];
+          });
+        }
       } else {
-
```
Comment (jaelim-ms) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: I'm so picky I guess. 😮  Blank line before comment.

File: MobileCenter/MobileCenter.xcodeproj/project.pbxproj (hunk 12 above)
Comment (TroubleMakerBen) [reply MobileCenter/MobileCenter.xcodeproj/project.pbxproj]: When you built the SDK using Xcode 9 and then drag'n'drop the SDK in an Xcode 8 project, you end up with a linker error that says `ld: framework not found FileProvider for architecture x86_64` which I don't understand why it happens, but I have been able to repro this again and again.

**Comment by TroubleMakerBen** [comment]
This works (tested distributing SasquatchObjC). Unit tests pass using Xcode 8, but not using Xcode 9 – this is not related to this change.
I've created a separate story for this that needs to be addressed.

Let me know what you think.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 14]
```diff
@@ -556,13 +599,12 @@ - (BOOL)handleUpdate:(MSReleaseDetails *)details {
...
 
 - (BOOL)checkForUpdatesAllowed {
-
-  // Check if we are not in AppStore or TestFlight environments.
-  BOOL environmentOkay = [MSUtility currentAppEnvironment] == MSEnvironmentOther;
-
-  // Check if a debugger is attached.
-  BOOL noDebuggerAttached = ![MSMobileCenter isDebuggerAttached];
-  return environmentOkay && noDebuggerAttached;
+  //  // Check if we are not in AppStore or TestFlight environments.
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Formatting issue here, should revert.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistributePrivate.h [hunk 15]
```diff
@@ -1,8 +1,21 @@
...
+#pragma clang diagnostic ignored "-Wreserved-id-macro"
+#ifndef __IPHONE_11_0
+#define __IPHONE_11_0    110000
+#endif
+#pragma clang diagnostic pop
+
 #import "MSAlertController.h"
 #import "MSUIAppDelegate.h"
 #import "MSDistribute.h"
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistributePrivate.h]: Any reasons for this?

Comment (clpolet) [reply MobileCenterDistribute/MobileCenterDistribute/MSDistributePrivate.h]: My 2 cents is SafariServices is weak linked so we should not have the need for this.

File: MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m [hunk 16]
```diff
@@ -446,7 +458,38 @@ - (nullable NSURL *)buildTokenRequestURLWithAppSecret:(NSString *)appSecret rele
   return components.URL;
 }
 
-- (void)openURLInEmbeddedSafari:(NSURL *)url fromClass:(Class)clazz {
+#if __IPHONE_OS_VERSION_MAX_ALLOWED >= __IPHONE_11_0
```
Comment (clpolet) [inline MobileCenterDistribute/MobileCenterDistribute/MSDistribute.m]: Couple of empty lines without comments :)

File: MobileCenter/MobileCenter.xcodeproj/project.pbxproj (hunk 12 above)
Comment (jaelim-ms) [reply MobileCenter/MobileCenter.xcodeproj/project.pbxproj]: Then why does it only matter on MobileCenter project?
